src/codecs/ico/encoder.rs

```diff
@@ -54,6 +55,48 @@ impl<W: Write> IcoEncoder<W> {
         self.w.write_all(&image_data)?;
         Ok(())
     }
+
+    /// Takes an already encoded PNG or BMP image and encodes it into an ICO.
+    ///
+    /// The dimensions of the image must be between 1 and 256 (inclusive) or
+    /// an error will be returned.
+    pub fn write_pre_encoded_image(self, encoded_image: &[u8]) -> ImageResult<()> {
+        let (width, height, color_type) = if encoded_image.starts_with(&PNG_SIGNATURE) {
+            let decoder = PngDecoder::new(Cursor::new(encoded_image))?;
+
+            let (width, height) = decoder.dimensions();
+            let color_type = decoder.color_type();
+            (width, height, color_type)
+        } else {
+            let decoder = BmpDecoder::new(Cursor::new(encoded_image))?;
+
+            let (width, height) = decoder.dimensions();
+            let color_type = decoder.color_type();
+            (width, height, color_type)
+        };
```

Comment:
This puts it in a weird spot where we decode _some_ of the data but do not fully validate it either. Could we not expose the `pre_encoded_image_impl` method directly?

Reply:
Sure, this is here only to make it harder to declare the wrong image dimensions or color type.